### PR TITLE
Fixed couple of errors and added new tests

### LINKED ISSUE #321: Trainer Evaluate not working
Probably as a result of #294 

Issue might be that there is a `detach().clone()` missing somewhere. What ends up happening is that an underlying pointer reference is passes on which later becomes 0. For e.g.

```python
rewards = torch.zeros(10) # Assuming there are 10 envs
rewards[i] = self.env.rewards[i] # This might be incompatible. As a clone operation might be needed.
```

Similarly for states, actions, etc

### PR DESCRIPTION
**Pull request by sampreet-arthi**
Fixes #321 and also solves the definition error in `action_shape` and `obs_shape`. There was an error that let OnPolicy Agents only solve either CartPole-v0 or Pendulum-v0 due to the differently defined `action_shape`

### CODE REVIEW

**Comment by sampreet-arthi** [comment]
Also need to add more tests (especially for On Policy algos). This one should've been caught by the unit tests.
